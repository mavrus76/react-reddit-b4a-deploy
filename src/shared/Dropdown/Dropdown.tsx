import React, { useEffect, useState } from 'react';
import styles from './dropdown.css';
import { noop } from '../../utils/js/noop';

interface IDropdownProps {
  button: React.ReactNode;
  children: React.ReactNode;
  isOpen?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}

export function Dropdown({
  button,
  children,
  isOpen,
  onClose = noop,
  onOpen = noop,
}: IDropdownProps) {
  const [isDropdownOpen, setIsDropdownOpen] = useState(isOpen);
  useEffect(() => setIsDropdownOpen(isOpen), [isOpen]);
  useEffect(() => (isDropdownOpen ? onOpen() : onClose()), [isDropdownOpen]);

  const handleOpen = () => {
    if (isOpen !== undefined) {
      setIsDropdownOpen(!isDropdownOpen);
    }
  };

  return (
    <div className={styles.container}>
      <div
        role='button'
        tabIndex={0}
        onClick={handleOpen}
        onKeyDown={(event) => {
          if (event.key === 'Enter' || event.key === ' ') {
            handleOpen();
          }
        }}>
        {button}
      </div>
      {isDropdownOpen && (
        <div className={styles.listContainer}>
          <div
            role='menuitem'
            className={styles.list}
            onClick={() => setIsDropdownOpen(false)}
            onKeyDown={(event) => {
              if (event.key === 'Escape' || event.key === 'Tab') {
                setIsDropdownOpen(false);
              }
            }}
            tabIndex={0}>
            {children}
          </div>
        </div>
      )}
    </div>
  );
}

Dropdown.defaultProps = {
  isOpen: false,
  onOpen: noop,
  onClose: noop,
};
