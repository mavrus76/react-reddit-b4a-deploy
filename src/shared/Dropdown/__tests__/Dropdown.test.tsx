/**
 * @jest-environment jsdom
 */

import React from 'react';
// eslint-disable-next-line enzyme/no-shallow
import { shallow } from 'enzyme';
import { Dropdown } from '../Dropdown';

describe('Dropdown', () => {
  test('should render', () => {
    const wrapper = shallow(
      <Dropdown button={<button type='button' />}>
        <div />
      </Dropdown>,
    );
    // console.log(window);
    expect(wrapper).toBeDefined();
    expect(wrapper.find('div.container').isEmptyRender()).toBeFalsy();
  });

  test('should render (snapshot)', () => {
    const wrapper = shallow(
      <Dropdown button={<button type='button' />}>
        <div />
      </Dropdown>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
