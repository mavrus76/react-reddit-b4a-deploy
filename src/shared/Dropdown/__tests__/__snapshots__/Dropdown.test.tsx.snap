// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render (snapshot) 1`] = `
<div
  className="container"
>
  <div
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <button
      type="button"
    />
  </div>
</div>
`;
